File: sem4/mpp/project-java/networking/src/main/java/networking/proxy/AbstractJsonProxy.java
package networking.proxy;

import com.google.gson.Gson;
import networking.Request;
import networking.Response;

public abstract class AbstractJsonProxy extends AbstractProxy {
	private final Gson gsonFormatter;

	public AbstractJsonProxy(String host, int port) {
		super(host, port);
		gsonFormatter = new Gson();
	}

	protected void sendRequest(Request request) throws Exception {
		String reqLine = gsonFormatter.toJson(request);
		System.out.println("request sent " + reqLine);
		try {
			output.println(reqLine);
			output.flush();
		} catch (Exception e) {
			System.out.println(e);
			throw new Exception("Error sending object " + e);
		}
	}

	protected Response readResponse() throws Exception {
		String responseLine = input.readLine();
		System.out.println("response received " + responseLine);
		return gsonFormatter.fromJson(responseLine, Response.class);
	}
}

File: sem4/mpp/project-cs/networking/Proxy/AbstractJsonProxy.cs
using networking.Proxy;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace networking;

public abstract class AbstractJsonProxy : AbstractProxy {
    protected AbstractJsonProxy(string host, int port) : base(host, port) {
        JsonConvert.DefaultSettings = () => {
            var settings = new JsonSerializerSettings {
                NullValueHandling = NullValueHandling.Ignore
            };
            settings.Converters.Add(new StringEnumConverter());
            return settings;
        };
    }

    protected override Response ReadResponse() {
        try {
            Console.WriteLine("reading response");
            var read = streamReader.ReadLine();
            Console.WriteLine("read " + read);
            return JsonConvert.DeserializeObject<Response>(read);
        }
        catch (Exception e) {
            throw new Exception("Error reading response " + e);
        }
    }


    protected void SendRequest(Request request) {
        try {
            var json = JsonConvert.SerializeObject(request);
            Console.WriteLine("writing: " + json);
            streamWriter.WriteLine(json);
            streamWriter.Flush();
            Console.WriteLine("request sent");
        }
        catch (Exception e) {
            throw new Exception("Error sending object " + e);
        }
    }
}